cf/weights.py

```diff
@@ -0,0 +1,1903 @@
+                return False
+
+            raise ValueError(
+                "Multiple weights specifications for "
+                f"{f.constructs.domain_axis_identity(axis)!r} axis"
+            )
+
+        radians = Units("radians")
+        metres = Units("metres")
+
+        x = aux_X.bounds.data
+        y = aux_Y.bounds.data
+
+        if x.Units.equivalent(radians) and y.Units.equivalent(radians):
+            if not great_circle:
+                raise ValueError(
+                    "Must set great_circle=True to allow the derivation of "
+                    "area weights from spherical polygons composed from "
+                    "great circle segments."
+                )
+
+            if methods:
+                weights[(axis,)] = "linear spherical line"
+                return True
+
+            spherical = True
+            x.Units = radians
+        elif x.Units.equivalent(metres) and y.Units.equivalent(metres):
+            if methods:
+                weights[(axis,)] = "linear plane line"
+                return True
+
+            spherical = False
+        else:
+            return False
+
+        y.Units = x.Units
+
+        if ugrid:
+            lengths = cls._line_length_ugrid(f, x, y, spherical)
+        else:
+            lengths = cls._line_length_geometry(f, x, y, spherical)
+
+        if not measure:
+            lengths.override_units(Units("1"), inplace=True)
+        elif spherical:
+            r = f.radius(default=radius)
+            r = r.override_units(Units("1"))
+            lengths = lengths * r
+
+        weights[(axis,)] = lengths
+        weights_axes.add(axis)
+        return True
+
+    @classmethod
+    def _line_length_geometry(cls, f, x, y, spherical):
+        """Creates line-length weights for line geometries.
+
+        .. versionadded:: 3.16.0
+
+        :Parameters:
+
+            f: `Field`
+                The field for which the weights are being created.
+
+            x: `Data`
+                The X coordinates of the line nodes.
+
+            y: `Data`
+                The Y coordinates of the line nodes.
+
+            spherical: `bool`
+                True for spherical lines.
+
+        :Returns:
+
+            `Data`
+                The length of the geometry line cells, in units of
+                metres.
+
+        """
+        from .units import Units
+
+        if spherical:
+            all_lengths = cls.central_angles(f, x, y)
+            all_lengths.override_units(Units("m"), inplace=True)
+        else:
+            delta_x = x.diff(axis=-1)
+            delta_y = y.diff(axis=-1)
+            all_lengths = (delta_x**2 + delta_y**2) ** 0.5
+
+        # Sum the lengths of each part to get the total length of each
+        # cell
+        lengths = all_lengths.sum(axes=(-2, -1), squeeze=True)
+        return lengths
+
+    @classmethod
+    def _line_length_ugrid(cls, f, x, y, spherical):
+        """Creates line-length weights for line geometries.
```

Comment:
This has the same description as for `_line_length_geometries`, so ideally we could update it, or clarify in a new sentence below, how it is different.

Reply:
ba891a7357996c22b4d5f6f6daa9d21d943f75e3